### PR TITLE
show similar commands with an edit distance of 2 or less

### PR DESCRIPTION
**Pull request by jakemac53**
Fixes https://github.com/dart-lang/args/issues/201

Uses the [Optimal string alignment distance](https://en.wikipedia.org/wiki/Damerau%E2%80%93Levenshtein_distance#Optimal_string_alignment_distance) algorithm to compute edit distance, which allows for additions, removals, substitutions, and swaps - all as 1 cost operations.

Suggests any commands within an edit distance of <=2, ~although we should possibly make this configurable~ which is now configurable.

We should also consider whether this should be opt in - it is likely to break existing integration tests in some cases. ~We could also consider adding in an opt out mechanism.~ you can now opt out by setting the distance to zero.

### CODE REVIEW

**Comment by jakemac53** [comment]
cc @jonahwilliams do you have opinions on this from a flutter tool perspective?

**Comment by jakemac53** [comment]
cc @devoncarew @jwren for comment as well as this likely would be used for the `dart` cli (although it could be disabled)

**Comment by jonahwilliams** [comment]
I love it

**Comment by jakemac53** [comment]
One additional thing I realized here - this will currently suggest hidden commands. Should it do that?

**Comment by jwren** [comment]
@bkonyi for `dartdev` fyi

LGTM.  Thanks for referencing the algorithm in the source code for future reference.

Nit: what about input where the user uses capitalization accidentally in the command name?

> hidden commands

Not sure.  If we do, we should write out that it is a hidden command: Did you mean `cmd_name (hidden)`?

**Comment by jonahwilliams** [comment]
when commands are disabled in the flutter tool we normally don't even add them, so I don't think it will be an issue

**Comment by bkonyi** [comment]
Awesome! I see no issues with this.

**Comment by jakemac53** [comment]
> Nit: what about input where the user uses capitalization accidentally in the command name?

This should work fine for that - it wouldn't handle it specially so it would be an edit distance of 1 to correct.

File: lib/command_runner.dart [hunk 1]
```diff
@@ -147,6 +154,21 @@ class CommandRunner<T> {
     Command? command;
     var commandString = executableName;
 
+    // Returns commands similar to `name`, in sorted order.
```
Comment (munificent) [inline lib/command_runner.dart:157]: Does this need to be a local function? Can you make it a private method?

File: lib/command_runner.dart [hunk 2]
```diff
@@ -147,6 +154,21 @@ class CommandRunner<T> {
...
+    Iterable<Command<T>> similarCommands(String name) {
+      final distances = Expando<int>();
```
Comment (munificent) [inline lib/command_runner.dart:159]: Why is this an expando and not just:

```dart
var distances = <String, int>{};
```

?

(Also, we use `var` for all locals in args.)

File: lib/command_runner.dart [hunk 3]
```diff
@@ -158,13 +180,32 @@ class CommandRunner<T> {
...
           command.usageException('Missing subcommand for "$commandString".');
         } else {
+          var requested = argResults.rest[0];
+
+          // Build up a help message containing similar commands, if found.
+          var similar = StringBuffer();
+          if (suggestedCommandsMaxEditDistance > 0) {
+            var possible = similarCommands(requested);
+            if (possible.isNotEmpty) {
+              similar.write(' The most similar command');
```
Comment (munificent) [inline lib/command_runner.dart:191]: How about "Did you mean one of these?"

**Comment by jakemac53** [comment]
I went ahead and changed this to omit hidden commands for now - we could always add a configuration option to enable it later

File: lib/command_runner.dart [hunk 4]
```diff
@@ -77,7 +77,14 @@ class CommandRunner<T> {
   ArgParser get argParser => _argParser;
   final ArgParser _argParser;
 
-  CommandRunner(this.executableName, this.description, {int? usageLineLength})
+  /// The maximum edit distance allowed when suggesting possible intended
+  /// commands.
+  ///
+  /// Set to `0` in order to disable suggestions, defaults to `2`.
+  final int suggestedCommandsMaxEditDistance;
```
Comment (munificent) [inline lib/command_runner.dart:84]: We're in `CommandRunner` so need to say "command" twice. How about just `suggestionDistanceLimit`?

File: lib/command_runner.dart [hunk 5]
```diff
@@ -158,13 +180,32 @@ class CommandRunner<T> {
...
+              if (possible.length == 1) {
+                similar.writeln(' is:');
+              } else {
+                similar.writeln('s are:');
+              }
+              for (var command in possible) {
+                similar.writeln('  ${command.name}');
+              }
+            }
+          }
```
Comment (munificent) [inline lib/command_runner.dart:201]: Also, I would move all this into the aforementioned method. Just have a single method that gives you the suggestion text and keep this one relatively simple.

File: lib/command_runner.dart [hunk 6]
```diff
@@ -433,3 +474,41 @@ String _getCommandUsage(Map<String, Command> commands,
...
+  // Add a space in front to mimic indexing by 1 instead of 0.
+  from = ' $from';
+  to = ' $to';
+  var distances = [
+    for (var i = 0; i < from.length; i++)
+      [
+        for (var j = 0; j < to.length; j++)
+          if (i == 0) j else if (j == 0) i else 0,
+      ],
+  ];
```
Comment (munificent) [inline lib/command_runner.dart:494]: Nit: Insert a blank line here and below the return statement.

File: lib/command_runner.dart [hunk 7]
```diff
@@ -433,3 +474,41 @@ String _getCommandUsage(Map<String, Command> commands,
...
+  for (var i = 1; i < from.length; i++) {
+    for (var j = 1; j < to.length; j++) {
+      // Removals from `from`
```
Comment (munificent) [inline lib/command_runner.dart:497]: "." at the end of all of these comments.

**Review by munificent — CHANGES_REQUESTED** [review]
This is really cool. People are going to love it.

File: lib/command_runner.dart (hunk 3 above)
Comment (jakemac53) [reply lib/command_runner.dart:191]: This is the text that `git` uses, but I agree it feels a bit awkard. I will update it.

File: lib/command_runner.dart (hunk 1 above)
Comment (jakemac53) [reply lib/command_runner.dart:157]: It does read state from the surrounding scope. I could instead pass that as a parameter, but the local method makes it a bit cleaner and it isn't used outside this method?

File: lib/command_runner.dart (hunk 3 above)
Comment (jakemac53) [reply lib/command_runner.dart:191]: Updated, I also changed the formatting a bit, generally looks like this now:

```
Could not find a command named "$typo".

Did you mean one of these?
  long
```

Comment (jakemac53) [reply lib/command_runner.dart]: cc @jonahwilliams @bkonyi how do you feel about the new text?

File: lib/command_runner.dart (hunk 1 above)
Comment (munificent) [reply lib/command_runner.dart]: If it were me, I would pass in the stuff as a parameter. This is quite long for a local function.

File: lib/command_runner.dart [hunk 8]
```diff
@@ -492,24 +489,26 @@ int _editDistance(String from, String to) {
           if (i == 0) j else if (j == 0) i else 0,
       ],
   ];
+
   for (var i = 1; i < from.length; i++) {
     for (var j = 1; j < to.length; j++) {
-      // Removals from `from`
+      // Removals from `from`,
```
Comment (munificent) [inline lib/command_runner.dart:495]: These should be ".", not ",". :)

https://dart.dev/guides/language/effective-dart/documentation#do-format-comments-like-sentences

**Review by munificent — APPROVED** [review]

File: lib/command_runner.dart (hunk 8 above)
Comment (jakemac53) [reply lib/command_runner.dart:495]: lol whoops, I am typing with my left hand atm lol

**Review by munificent — APPROVED** [review]

**Pull request merged**